Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -11,17 +11,116 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
-/// Analyzes the types of variables.
-public struct AbstractInterpreter {
-    // States are kept in a stack to support conditional execution.
-    private var ifStack: [VariableMap<Type>] = []
-    private var stack = [VariableMap<Type>()]
+private struct InterpreterState {
+    // Save variable changes, supports
+    // 1. Nested states (function states, loop states)
+    // 2. Sibling states (if/else branches)
+    private var stateChangesStack = [[VariableMap<Type>()], [VariableMap<Type>()]]
     
     // The currently active state.
-    private var currentState: VariableMap<Type> {
-        return stack.last!
+    private var currentState = VariableMap<Type>()
+
+    public mutating func reset() {
+        stateChangesStack = [[VariableMap<Type>()], [VariableMap<Type>()]]
+        currentState = VariableMap<Type>()
+    }
+
+    /// Update current variable types and type changes
+    /// Used after block end when some states should be merged to parent
+    public mutating func mergeStates(typeChanges: inout [(Variable, Type)]) {
+        let states = stateChangesStack.removeLast()
+        // Active state of the previous states entry
+        let prevState = stateChangesStack.last!.last!
+        var numUpdatesPerVariable = VariableMap<Int>()
+        var newTypes = VariableMap<Type>()
+
+        for state in states {
+            for (v, t) in state {
+                // Skip variable types that are out of scope as we do not care about them anymore
+                guard t != .nothing else { continue }
+
+                if newTypes[v] == nil {
+                    newTypes[v] = t
+                    numUpdatesPerVariable[v] = 1
+                } else {
+                    newTypes[v]! |= t
+                    numUpdatesPerVariable[v]! += 1
+                }
+            }
+        }
+
+        for (v, c) in numUpdatesPerVariable {
+            // Skip updating:
+            // 1. Fallback type for grandchild (no record in prevState)
+            // 2. Local variables, which go out of scope (prevState is .nothing)
+            if prevState[v] == nil || prevState[v] == .nothing {
```

Comment:
When does the `prevState[v] == nil` case occur? Shouldn't `setType` ensure that the parent state always has the old type? Probably related: shouldn't this function also ensure that the new parent state has the previous type for things that changed, e.g. how is this handled:
```
let x = 42;
// diff: x = int
if (c1) {
    // diff: <empty> // but should it be int here?
    if (c2) {
        // diff: x = int (when setting x in the child)
        if (c3) {
            x = "foobar";
            // diff: x = string
        } else {}
        // diff: x = int | string
    } else {}
} else {}
// does x have the correct type here?
```

Reply:
well when you set type to parentState then later on parentState becomes current and sudden of all parent state does not have any type

Reply:
but yeah you're right I should probably use `setType` function